citrixitm/resource_citrixitm_dns_app.go

```diff
@@ -86,18 +86,22 @@ func withExistingResource(f ProcessAppFunc) func(*schema.ResourceData, interface
 }
 
 func read(id int, c *itm.Client, d *schema.ResourceData) error {
-	app, err := c.DnsApps.Get(id)
-	log.Printf("[DEBUG] Inside read; app: %#v", app)
+	app, err := c.DNSApps.Get(id)
 	if err != nil {
 		return fmt.Errorf("Error retrieving app: %s", err)
 	}
-	d.Set("name", app.Name)
-	d.Set("description", app.Description)
-	d.Set("fallback_cname", app.FallbackCname)
-	d.Set("fallback_ttl", app.FallbackTtl)
-	d.Set("app_data", app.AppData)
-	d.Set("cname", app.AppCname)
-	d.Set("version", app.Version)
+	if app.Enabled {
+		d.Set("name", app.Name)
+		d.Set("description", app.Description)
+		d.Set("fallback_cname", app.FallbackCname)
+		d.Set("fallback_ttl", app.FallbackTtl)
+		d.Set("app_data", app.AppData)
+		d.Set("cname", app.AppCname)
+		d.Set("version", app.Version)
+	} else {
+		log.Printf("The app is disabled. This likely means that it was deleted outside of Terraform.")
```

Comment:
what is the workflow after this error message is logged? i.e. will the same error message show up the next time this is run or will the user need to override some config value to allow this deleted app to be recreated?

Reply:
I can make that message more clear. The user doesn't need to do anything because from their perspective, once an app is deleted in the Portal, it's simply gone and can't be recovered. In conversation with John and Steve, it was determined that Terraform should offer to recreate the app in this case.

The workflow depends on what action the user is performing at the time. `terraform plan` doesn't have any side effects and continues to log this message every time it is run until the user runs `terraform apply`, which also logs this message (both commands cross this code path). But if the user confirms the change, `terraform apply` goes on to recreate the app and the Terraform state is assigned the new app's ID.

I should point out that examining Terraform logs is not something the user typically does. Logging needs to be explicitly enabled. It's mostly for developers and my impression is that they tend to be extremely verbose. Plugins can preface their messages with text like "[DEBUG]" to limit how much Terraform outputs depending on logging related settings. This particular message is a good warning level message so I think I'll add that at the same time that I rework the message itself.

However, the logs are still not how the typical user would learn about why the app is being recreated. The typical Terraform user would scratch their head for a bit and then come to the conclusion that a resource they were managing with Terraform must have gotten clobbered externally, which is a big Terraform no-no.

Reply:
Now says,

> The app is disabled. This likely means that it was deleted outside of Terraform. 'terraform apply' will recreate the app if you approve. If you don't wish Terraform to continue prompting about it, then you may want to remove its configuration.

Reply:
that reads better. FWIW, the portal now allows deleted apps to be brought back from the dead (kinda orthogonal but worth mentioning any way)